Upgrade go perun v0.6.0

matthiasgeihs opened the pull request:
This changes upgrades from go-perun v0.4.0 to v0.6.0. This involves changes to the funding, watching, settling logic.

In addition, some other changes are included:
- User logging is revised for better user guidance
- The user receiving a channel proposal now must accept the proposal before the channel is opened

matthiasgeihs committed:
:children_crossing: Additional logging for user guidance

matthiasgeihs committed:
:children_crossing: Make port numbers consistent with node start order

matthiasgeihs committed:
:children_crossing: [CLI] Hint at help command and do not complain about empty command

matthiasgeihs committed:
:children_crossing: Revise info command documentation

matthiasgeihs committed:
:children_crossing: Connect on open

matthiasgeihs committed:
:sparkles: Benchmark with amount

ggwpez commented on cmd/demo/node.go:
```diff
@@ -128,16 +132,44 @@ func (n *node) setupChannel(ch *client.Channel) {
+
+	if _, ok := e.(*channel.ConcludedEvent); ok {
+		fmt.Printf("\n🎭 Concluded event\n")
+		func() {
```
I think this wrapper function is not needed.

matthiasgeihs replied:
it's not necessary now, but it could be useful if somebody extends the function below. you want to unlock after settle...

ggwpez commented on cmd/demo/node.go:
```diff
@@ -219,9 +256,8 @@ func (n *node) HandleProposal(req *client.ChannelProposal, res *client.ProposalR
-	if _, err := res.Accept(ctx, client.ProposalAcc{
-		Participant: n.offChain.Address(),
-	}); err != nil {
+	a := req.Accept(n.offChain.Address(), client.WithNonce(nonce()))
```
You can use `client.WithRandomNonce()` here and in the other use-case and then remove the `nonce` function.

matthiasgeihs replied:
done

ggwpez replied:
There is still another use of `client.WithRandomNonce` below.

matthiasgeihs replied:
hm? not sure what you mean.

ggwpez replied:
There is a call to `WithRandomNonce` in line 308 as argument for `NewLedgerChannelProposal` which is not needed.

ggwpez commented on cmd/demo/node.go:
```diff
@@ -230,13 +266,7 @@ func (n *node) HandleProposal(req *client.ChannelProposal, res *client.ProposalR
-	n.mtx.Lock()
-	defer n.mtx.Unlock()
-	n.settle(p)
+	// Nothing to do here. Channel closing is handled in on-chain event handler.
```
Why keep the empty function?

matthiasgeihs replied:
removed, together with the onFinal field on the paymentChannel object.

ggwpez commented on cmd/demo/node.go:
```diff
@@ -253,14 +283,17 @@ func (n *node) Open(args []string) error {
+		n.offChain.Address(),
+		initBals,
+		[]wire.Address{n.onChain.Address(), peer.perunID},
+		client.WithNonce(nonce()),
```
`NewLedgerChannelProposal` should use a random nonce when none is passed, so it should be ok to omit this.

matthiasgeihs replied:
done

ggwpez commented on cmd/demo/node_setup.go:
```diff
@@ -33,15 +37,18 @@ import (
 var (
 	backend         *node
 	ethereumBackend *ethclient.Client
+	app             = &payment.App{Addr: &ewallet.Address{}}
```
Instead of using the `[payment] App` here we should use the `NoApp` and do the check in the application logic. @seb  

 
After looking at it, i realized the the check in `PaymentChannel.Handle` is missing :see_no_evil: (will prepare a patch).

matthiasgeihs replied:
which application logic?

matthiasgeihs replied:
i am assuming by application logic you meant the node logic. extended function `paymentChannel.Handle` accordingly.

ggwpez replied:
So you just copied the `ValidTransition` function from the `payment.App` here, you can actually reuse it with
```go
var app payment.App
app.ValidTransition(nil, ch.lastState, update.State, update.ActorIdx)
```
what do you think?

matthiasgeihs replied:
I think it's a bit weird. If we say, it doesn't make sense to have the payment logic in an app, then we should remove the payment app completely.

ggwpez replied:
Ok, i added a refactor issue in go-perun for the `ValidTransition` function.

ggwpez commented on cmd/demo/node_setup.go:
```diff
@@ -33,15 +37,18 @@ import (
-	payment.SetAppDef(appDef)
+	// The payment app currently has no on-chain logic. Therefore, we do not
+	// need to deploy the app on-chain.
+	channel.RegisterApp(app)
```
Without an app, we do not need this either :tada:

matthiasgeihs replied:
so you want me to get rid of the app completely? and instead check in the HandleUpdate, that the state change is fine?

matthiasgeihs replied:
after thinking about it, made sense to me. done that.

ggwpez commented on cmd/demo/node_setup.go:
```diff
@@ -50,6 +57,7 @@ func Setup() {
 	if backend, err = newNode(); err != nil {
 		log.WithError(err).Fatalln("Could not initialize node.")
 	}
+
```
Accident?

matthiasgeihs replied:
yes

ggwpez commented on cmd/demo/node_setup.go:
```diff
@@ -120,19 +129,19 @@ func (n *node) setupContracts() error {
 	case contractSetupOptionDeploy:
-		if adjAddr, err = deployAdjudicator(n.cb); err == nil { // deploy adjudicator
-			assAddr, err = deployAssetHolder(n.cb, adjAddr) // deploy asset holder
+		if adjAddr, err = n.deployAdjudicator(n.cb); err == nil { // deploy adjudicator
```
If you make this a member function you can remove the `n.cb` argument, same for `deployAssetHolder`.
It would then also make more sense to directly set `n.adjAddr` and `assetAddr` instead of returning it.

matthiasgeihs replied:
background: i made it a member function because deployment now requires an account and that the node account is stored with the node struct.

Now that I view my code once again, I think it is cleaner to not make it a member function, but just also supply the account as a parameter. less change necessary.

ggwpez replied:
I also think its better to not have them as member functions.

ggwpez commented on cmd/demo/node_setup.go:
```diff
@@ -147,11 +156,15 @@ func (n *node) setupContracts() error {
 	recvAddr := ewallet.AsEthAddr(n.onChain.Address())
-	n.adjudicator = echannel.NewAdjudicator(n.cb, n.adjAddr, recvAddr)
-	n.funder = echannel.NewETHFunder(n.cb, n.assetAddr)
+	acc := n.onChain.Account
```
I think this does not increase readability.

matthiasgeihs replied:
ok

ggwpez commented on cmd/demo/node_setup.go:
```diff
@@ -199,26 +212,26 @@ func validateAssetHolder(cb echannel.ContractBackend, adjAddr common.Address) (c
 
 // importAccount is a helper method to import secret keys until we have the ethereum wallet done.
-func importAccount(secret string) (*ewallet.Wallet, *ewallet.Account, error) {
+func importAccount(secret string) (*pkeystore.Wallet, *pkeystore.Account, error) {
```
Since we now have a proper mnemonic wallet we should use that instead, i will add an issue.

matthiasgeihs replied:
ok

ggwpez replied:
Its here https://github.com/perun-network/perun-eth-demo/issues/27

ggwpez commented on cmd/demo/validator.go:
```diff
@@ -84,15 +86,11 @@ func strToAddress(str string) (wallet.Address, error) {
-	val, err := srand.Int(srand.Reader, max)
-	if err != nil {
-		log.Panic("Could not create nonce")
+func nonce() (share client.NonceShare) {
```
Should be superfluous now.

matthiasgeihs replied:
yes, removed

ggwpez commented on cmd/demo/commands.go:
```diff
@@ -107,8 +107,8 @@ func Executor(in string) error {
 	}
-	if command != "" {
-		return errors.Errorf("Unknown command: %s", command)
+	if len(command) > 0 {
```
Could you maybe squash this with `8ddb2fda77de80c18e9468c5fbbcce1b5544af2d`?

matthiasgeihs replied:
ok

ggwpez commented on cmd/demo/node.go:
```diff
@@ -296,6 +302,9 @@ func (n *node) Open(args []string) error {
 		return errors.WithMessage(err, "creating channel proposal")
 	}
 
+	alias := args[0]
```
You like doing this dont you :laughing: ?

matthiasgeihs replied:
I didn't realize that there was already a variable for args[0] above :D (ideally, I think, the variables should be parsed before calling open, but anyways...)

ggwpez replied:
They are currently parsed but only to verify that they are of the correct type.  
Would be nicer to have them passed as correct types but then the setup in `commands.go` would not be so compact.

ggwpez commented on cmd/demo/commands.go:
```diff
@@ -58,7 +58,7 @@ func init() {
 			"info",
 			nil,
-			"Print the current open state channels.",
+			"Print information about funds, peers, and channels.",
```
You could squash this commit into `a2c98f4a9f9a6c29aec6003bceb10cd0c5a603a0` if you want, since both update logging/printed documentation.

matthiasgeihs replied:
Hm, I rather keep it separate here, as it touches different aspects.

ggwpez commented on README.md:
````diff
@@ -61,10 +61,15 @@ Once both CLIs are running, e.g. in Alice's terminal, connect to bob with
 ```
+In Alice's terminal, accept the appearing channel proposal.
+```
+🔁 Incoming channel proposal from alice with funding [My: 10 Ξ, Peer: 10 Ξ].
````
Nice idea :sunglasses:

matthiasgeihs replied:
thx!

ggwpez commented on cmd/demo/node.go:
```diff
@@ -245,7 +243,11 @@ func (n *node) HandleProposal(prop client.ChannelProposal, res *client.ProposalR
 	if p == nil {
 		if cfg == nil {
-			res.Reject(ctx, "Unknown identity")
+			func() {
```
Did you do this `func` so that the context is always cancelled?  
The function should return quickly anyway i think.

matthiasgeihs replied:
The idea was to ensure that the context is shortly canceled after function `Reject` returns. Doesn't seem to be necessary, right? --> removed

ggwpez replied:
Yes it should be ok without.

ggwpez commented on cmd/demo/node.go:
```diff
@@ -245,7 +243,11 @@ func (n *node) HandleProposal(prop client.ChannelProposal, res *client.ProposalR
+			func() {
+				ctx, cancel := context.WithTimeout(context.Background(), config.Node.HandleTimeout)
+				defer cancel()
+				res.Reject(ctx, "Unknown identity")
```
Logging or handling of the returned error missing here.

matthiasgeihs replied:
logging the error now

ggwpez commented on cmd/demo/node.go:
```diff
@@ -258,15 +260,29 @@ func (n *node) HandleProposal(prop client.ChannelProposal, res *client.ProposalR
+			a := req.Accept(n.offChain.Address(), client.WithNonce(nonce()))
+			if _, err := res.Accept(ctx, a); err != nil {
+				n.log.Error(errors.WithMessage(err, "accepting channel proposal"))
+				return
```
You can remove the returns here and below.

matthiasgeihs replied:
you are right, currently the functionality would not change if we remove `return`. however, logically, the function should always return in case of error and I like to be explicit about that. for example, if some other develop later adds code below and is not aware that the function already failed, this could be problematic.

ggwpez commented on cmd/demo/node.go:
```diff
@@ -347,7 +363,11 @@ func (n *node) Close(args []string) error {
+	if err := n.settle(peer); err != nil {
+		return errors.WithMessage(err, "settling")
+	}
+	fmt.Printf("\r🏁 Settled channel with %s.\n", peer.alias)
```
The `\r` is MacOS specific, right? Why is it needed here btw? Why not use the `PrintfAsync` here?

matthiasgeihs replied:
No, `r` is not MacOS specific. This character tells the printer (e.g., the terminal) to move the cursor to the beginning of the line (see [Carriage return](https://en.wikipedia.org/wiki/Carriage_return)). I use it for overwriting the prefix `>` when printing.

I don't use `PrintfAsync` because `Close` is not triggered by an async call (e.g., a channel event). However, even though `Close` is not called asynchronously, the settle call contains an async print, which adds `>` to the next line. Hence, the `\r` is still necessary to remove.

This whole trouble comes from the fact that we cannot tell the go-prompt to clean print on an async event. To have a clean looking console without intermittent `>` prefix characters, we therefore use `PrintfAsync`, which first overwrites the existing `>`, then prints the message, and then adds the `>` again.

ggwpez replied:
I see, makes sense.

ggwpez commented on cmd/demo/commands.go:
```diff
@@ -85,9 +88,46 @@ func init() {
-// Executor interprets commands entered by the user.
-// Gets called by Cobra, but could also be used for emulating user input.
-func Executor(in string) error {
+var asyncInputRequest chan struct{} = make(chan struct{})
```
Instead of having three channels here you could also go with one `var prompts = make(chan func(string), 1)`.  
Let the `AddInput` function be synchronous again (could also reorder commands otherwise) and have `Prompt` be non-blocking.

matthiasgeihs replied:
Works!

ggwpez commented on cmd/demo/commands.go:
```diff
@@ -40,7 +40,7 @@ func init() {
 		}, {
 			"open",
-			[]argument{{"Peer", valPeer}, {"Our Balance", valBal}, {"Their Balance", valBal}},
+			[]argument{{"Peer", valAlias}, {"Our Balance", valBal}, {"Their Balance", valBal}},
```
You renamed it here but not in the definition :man_shrugging:

matthiasgeihs replied:
hm?

ggwpez replied:
Ok no idea what i meant :laughing: I probably did not see the `valAlias` function.

ggwpez commented on cmd/demo/node.go:
```diff
@@ -294,9 +294,16 @@ func (n *node) handleFinal(p *peer) {
 	if peer == nil {
-		return errors.Errorf("peer not found %s", args[0])
+		// try to connect to peer
+		n.mtx.Unlock()
```
It would be better to have a private `connect` function which is then called from here and from `Connect` instead of having this lock/unlock thing here.

matthiasgeihs replied:
sure

ggwpez commented on README.md:
````diff
@@ -81,10 +81,9 @@ You may always check the current status with command `info`.
 You can also run a performance benchmark with command
 ```
-> benchmark alice 1000
+> benchmark alice 10 100
````
When we add an amount to the benchmark command the console of both parties is spammed with `Received balance update` messages, where is the benefit of having an amount here?

matthiasgeihs replied:
It's how we did it in Erdstall and I think it's better for a demo, because you see that actually something is happening. Additionally, you can verify that the the micro transactions had an impact on your on-chain balance after you close the channel.

ggwpez replied:
Ok sounds reasonable.

ggwpez commented on cmd/demo/benchmark.go:
```diff
@@ -53,20 +52,23 @@ func (n *node) Benchmark(args []string) error {
 	defer n.mtx.Unlock()
 	peer := n.peers[args[0]]
-	x, _ := strconv.Atoi(args[1])
+	totalAmountEth, _ := strconv.Atoi(args[1])
```
This is used as float below but you parse it as int here (if you change it here, you also need to use the other validator in the command definition).

matthiasgeihs replied:
This was somewhat intentional, but allowing floats is also fine with me.

ggwpez replied:
I dont mind for now, will close it.

matthiasgeihs committed:
:arrow_up: [go.mod] Upgrade dependencies

Upgrade to go-perun v0.6.0. This includes updating funding, watching, and settling logic.

matthiasgeihs committed:
:children_crossing: Revise contract validation console output

matthiasgeihs committed:
:children_crossing: :sparkles: Require channel opening approval from user

Includes adding support for reading user input during event handling.

ggwpez reviewed (changes requested):
<s>I just looked at the CI and saw that the Unit tests always report no test files, so they dont seem to run.</s>
